adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+                # a) If word has never been seen before, go through initialization step
+                self.initialization_step(word, observed_perception_graph)
+            else:
+                # b) If word has been seen before, test the recorded hypothesis
+                if self.test_hypothesis_step(word, observed_perception_graph):
+                    # If hypothesis is confirmed, do nothing
+                    continue
+                else:
+                    # If hypothesis is not confirmed, treat word as an unseen word, and go
+                    # through the initialization step again (to randomly select meaning)
+                    self.initialization_step(word, observed_perception_graph)
+            if self._log_word_hypotheses_to:
+                self._log_hypotheses(word)
+
+    def initialization_step(self, word: str, observed_perception_graph: PerceptionGraph):
+        # If the word has never been seen before OR word does not match any
+        # 'object' in the current frame, learn a new hypothesis/pattern
+        # generated as a pattern graph from the perception.
+        meanings = [
+            PerceptionGraphPattern.from_graph(object_).perception_graph_pattern
+            for object_ in self.get_objects_from_perception(observed_perception_graph)
+        ]
+
+        #  Pick a random meaning for our hypothesis (Independent of any prior data)
+        pattern_hypothesis = self._rng.choice(meanings)
+
+        if self._graph_logger:
+            self._graph_logger.log_graph(
+                pattern_hypothesis,
+                logging.INFO,
+                "Initializing (randomly chosen) meaning for %s ",
+                word,
+            )
+
+        self._words_to_hypotheses[word] = pattern_hypothesis
+
+    # Test Hypothesis
+    def test_hypothesis_step(
+        self, word: str, observed_perception_graph: PerceptionGraph
+    ) -> bool:
+        # Compare existing hypothesis with observation
+        # Return whether
+        hypotheses_for_word = self._words_to_hypotheses[word]
+        logging.info("Current hypothesis is %s", abs(hash((word, hypotheses_for_word))))
+        self.debug_counter += 1
+        # If the leading hypothesis sufficiently matches the observation, reinforce it
+        # To do, we check how much of the leading pattern hypothesis matches the perception
+        partial_match = self._compute_match_ratio(
+            hypotheses_for_word, observed_perception_graph
+        )
+
+        hypothesis_is_confirmed = partial_match.matched_exactly()
+        # If the hypothesis is confirmed, we leave it as it is
+        if hypothesis_is_confirmed and partial_match.matching_subgraph:
+            logging.info("Hypothesis is confirmed")
+        # If the hypothesis is disconfirmed
+        else:
+            logging.info(
+                f"Hypothesis is disconfirmed (Match ratio: {partial_match.match_ratio()}"
+            )
+        return hypothesis_is_confirmed
```

Comment:
Can you add a parameter (e.g minimum_match_ratio) into the learner such that a hypothesis is confirmed if the match ratio >= the minimum match ratio? The learner as is might never successfully verify because of minor variations in the perceptual representations.

Reply:
Added. The default minimum_match_ratio has been set to 0.8; with the test parameters using 0.85